File: js/src/validate.js
/**
 * Local validation.
 *
 * Every rule here mirrors something the NetCrunch receiver discards *silently* —
 * an empty status value, a status key it reserves, an event with no message. A
 * library that forwarded those would lose data with nothing raised at either end,
 * so each is rejected at the call site instead, where the stack trace points at
 * the code that got it wrong.
 *
 * See spec/v1.md sections 3 to 5.
 */

export const MAX_STATUS_KEY_LENGTH = 500;

function describe(value) {
  if (value === null) return "null";
  if (value === undefined) return "undefined";
  if (value === "") return "an empty string";
  if (typeof value === "string" && value.trim() === "") return "a blank string";
  return typeof value;
}

export function assertCounterPath(object, counter) {
  if (typeof object !== "string" || object.trim() === "") {
    throw new TypeError(`Counter object is required and must be a non-empty string (got ${describe(object)}).`);
  }
  if (typeof counter !== "string" || counter.trim() === "") {
    throw new TypeError(`Counter name is required and must be a non-empty string (got ${describe(counter)}).`);
  }
}

export function assertCounterInstance(instance) {
  if (instance === undefined || instance === null) return;
  if (typeof instance !== "string") {
    throw new TypeError(`Counter instance must be a string (got ${describe(instance)}).`);
  }
}

export function assertCounterValue(value) {
  if (typeof value !== "number") {
    throw new TypeError(`Counter value must be a number (got ${describe(value)}). Use status() for text.`);
  }
  // NaN and Infinity have no JSON representation — JSON.stringify turns them into
  // null, which the receiver would take as a legitimate value.
  if (!Number.isFinite(value)) {
    throw new RangeError(`Counter value must be finite (got ${value}).`);
  }
}

export function assertStatusKey(key) {
  if (typeof key !== "string" || key.trim() === "") {
    throw new TypeError(`Status key is required and must be a non-empty string (got ${describe(key)}).`);
  }
  if (key.startsWith("@")) {
    throw new TypeError(`Status key "${key}" is reserved — NetCrunch uses the "@" prefix internally.`);
  }
  if (key.length > MAX_STATUS_KEY_LENGTH) {
    throw new RangeError(`Status key is ${key.length} characters; NetCrunch truncates at ${MAX_STATUS_KEY_LENGTH}.`);
  }
}

export function assertStatusValue(value) {
  if (typeof value !== "string") {
    throw new TypeError(`Status value must be a string (got ${describe(value)}). Use counter() for numbers.`);
  }
  if (value === "") {
    throw new TypeError("Status value must not be empty — NetCrunch discards empty statuses without reporting it.");
  }
}

/** Beyond this the receiver slices arrays without telling anyone. */
export const MAX_DATA_ENTRIES = 1024;

/** Type to the members that carry its payload. Also the set of accepted types. */
export const DATA_TYPE_MEMBERS = {
  table: ["columns", "rows"],
  "time-series": ["timestamps", "values"],
  category: ["categories", "values"],
};

const DATA_TYPES = DATA_TYPE_MEMBERS;

export function assertDataObject(id, type, details) {
  if (typeof id !== "string" || id.trim() === "") {
    throw new TypeError(`Data object id is required and must be a non-empty string (got ${describe(id)}).`);
  }

  if (type === "internal") {
    throw new TypeError('The "internal" data object type is reserved for NetCrunch\'s own sensors.');
  }

  const required = DATA_TYPES[type];
  if (required === undefined) {
    const known = Object.keys(DATA_TYPES).join(", ");
    throw new TypeError(
      `Unknown data object type "${type}". NetCrunch discards these with only a server-side warning. Use one of: ${known}.`
    );
  }

  for (const member of required) {
    if (!Array.isArray(details[member])) {
      throw new TypeError(`A ${type} data object requires "${member}" to be an array.`);
    }
    if (details[member].length > MAX_DATA_ENTRIES) {
      throw new RangeError(
        `"${member}" has ${details[member].length} entries; NetCrunch truncates at ${MAX_DATA_ENTRIES} without reporting it.`
      );
    }
  }

  // Ragged parallel arrays are the dangerous case: nothing errors anywhere, and
  // the chart quietly plots the wrong thing.
  if (type === "table") {
    const width = details.columns.length;
    details.rows.forEach((row, index) => {
      if (!Array.isArray(row)) {
        throw new TypeError(`Table row ${index} must be an array of cells.`);
      }
      if (row.length !== width) {
        throw new RangeError(
          `Table row ${index} has ${row.length} cells but there are ${width} columns.`
        );
      }
    });
  } else {
    const [left, right] = required;
    if (details[left].length !== details[right].length) {
      throw new RangeError(
        `"${left}" has ${details[left].length} entries but "${right}" has ${details[right].length}; they must match.`
      );
    }
  }
}

export function assertEventMessage(message) {
  if (typeof message !== "string") {
    throw new TypeError(`Event message must be a string (got ${describe(message)}).`);
  }
  if (message.trim() === "") {
    throw new TypeError("Event message must not be empty — NetCrunch discards such events without reporting it.");
  }
}

File: go/validate.go
package telemetry

import (
	"fmt"
	"math"
	"reflect"
	"strings"
)

const (
	maxStatusKeyLength = 500

	// Beyond this the receiver slices arrays without telling anyone.
	maxDataEntries = 1024
)

// dataTypeMembers maps a data object type to the members carrying its payload.
// It is also the set of accepted types.
var dataTypeMembers = map[string][]string{
	"table":       {"columns", "rows"},
	"time-series": {"timestamps", "values"},
	"category":    {"categories", "values"},
}

// Validation here mirrors what the NetCrunch receiver discards *silently* — an
// empty status value, a key it reserves, an event with no message. A library
// that forwarded those would lose data with nothing raised at either end, so
// each is rejected at the call site, where the caller can still do something
// about it.
//
// Go's type system removes several of the checks the other implementations need:
// a counter value cannot be a string, and a status value cannot be a number,
// because the signatures do not allow it.

func validateCounterPath(object, counter string) error {
	if strings.TrimSpace(object) == "" {
		return fmt.Errorf("counter object is required and must not be empty")
	}
	if strings.TrimSpace(counter) == "" {
		return fmt.Errorf("counter name is required and must not be empty")
	}
	return nil
}

func validateCounterValue(value float64) error {
	// NaN and Inf have no JSON representation; encoding/json refuses them, and a
	// silent zero would be worse than the error.
	if math.IsNaN(value) || math.IsInf(value, 0) {
		return fmt.Errorf("counter value must be finite, got %v", value)
	}
	return nil
}

func validateStatusKey(key string) error {
	if strings.TrimSpace(key) == "" {
		return fmt.Errorf("status key is required and must not be empty")
	}
	if strings.HasPrefix(key, "@") {
		return fmt.Errorf("status key %q is reserved — NetCrunch uses the \"@\" prefix internally", key)
	}
	if len(key) > maxStatusKeyLength {
		return fmt.Errorf("status key is %d characters; NetCrunch truncates at %d", len(key), maxStatusKeyLength)
	}
	return nil
}

func validateStatusValue(value string) error {
	if value == "" {
		return fmt.Errorf("status value must not be empty — NetCrunch discards empty statuses without reporting it")
	}
	return nil
}

func validateEventMessage(message string) error {
	if strings.TrimSpace(message) == "" {
		return fmt.Errorf("event message must not be empty — NetCrunch discards such events without reporting it")
	}
	return nil
}

// validateDataObject checks a data object against spec/v1.md section 6.
func validateDataObject(id, objectType string, members map[string]any) error {
	if strings.TrimSpace(id) == "" {
		return fmt.Errorf("data object id is required and must not be empty")
	}
	if objectType == "internal" {
		return fmt.Errorf(`the "internal" data object type is reserved for NetCrunch's own sensors`)
	}

	required, ok := dataTypeMembers[objectType]
	if !ok {
		return fmt.Errorf(
			"unknown data object type %q — NetCrunch discards these with only a server-side warning; use table, time-series or category",
			objectType,
		)
	}

	lengths := make(map[string]int, len(required))
	for _, member := range required {
		length, ok := memberLength(members[member])
		if !ok {
			return fmt.Errorf("a %s data object requires %q to be an array", objectType, member)
		}
		if length > maxDataEntries {
			return fmt.Errorf(
				"%q has %d entries; NetCrunch truncates at %d without reporting it",
				member, length, maxDataEntries,
			)
		}
		lengths[member] = length
	}

	// Ragged parallel arrays are the dangerous case: nothing errors anywhere and
	// the chart quietly plots the wrong thing.
	if objectType == "table" {
		width := lengths["columns"]
		rows, _ := members["rows"].([]any)
		for i, row := range rows {
			cells, ok := row.([]any)
			if !ok {
				return fmt.Errorf("table row %d must be an array of cells", i)
			}
			if len(cells) != width {
				return fmt.Errorf("table row %d has %d cells but there are %d columns", i, len(cells), width)
			}
		}
		return nil
	}

	left, right := required[0], required[1]
	if lengths[left] != lengths[right] {
		return fmt.Errorf("%q has %d entries but %q has %d; they must match", left, lengths[left], right, lengths[right])
	}
	return nil
}

// isNil reports whether value is nil, including a typed nil held in an interface
// — which `value == nil` alone does not catch.
func isNil(value any) bool {
	if value == nil {
		return true
	}
	switch reflected := reflect.ValueOf(value); reflected.Kind() {
	case reflect.Chan, reflect.Func, reflect.Interface, reflect.Map, reflect.Pointer, reflect.Slice:
		return reflected.IsNil()
	default:
		return false
	}
}

// memberLength reports the length of a data object member, accepting the several
// slice types the typed constructors and the generic Data method both produce.
func memberLength(value any) (int, bool) {
	switch typed := value.(type) {
	case nil:
		return 0, false
	case []any:
		return len(typed), true
	case []string:
		return len(typed), true
	case []float64:
		return len(typed), true
	case []int64:
		return len(typed), true
	case [][]any:
		return len(typed), true
	default:
		return 0, false
	}
}
